### PR TITLE
util: Check/remove empty dat files

### PR DESCRIPTION
**Pull request by yeominstall**
remove automatically created dat files for each online CPUs like ‘perf-cpuXX.dat’ and 'kernel-cpuXX.dat’ if it doesn’t have no content in it.

Fixed: #893

### CODE REVIEW

**Comment by yeominstall** [comment]
```
/home/yminn/uftrace/utils/perf.c: In function ‘finish_perf_record’:
/home/yminn/uftrace/utils/perf.c:154:9: warning: implicit declaration of function ‘stat’; did you mean ‘strcat’? [-Wimplicit-function-declaration]
    if (!stat(filename, &stbuf) && stbuf.st_size == 0) {
         ^~~~
         strcat
```
I think it works fine, but it occurs a warning message when I build it.

**Comment by honggyukim** [comment]
You can add the following include statements to hide warnings. You can find the proper header files in the man page for `stat`.
```diff
diff --git a/utils/perf.c b/utils/perf.c
index 9d5e086..7950bd0 100644
--- a/utils/perf.c
+++ b/utils/perf.c
@@ -7,6 +7,8 @@
 #include <byteswap.h>
 #include <sys/mman.h>
 #include <sys/syscall.h>
+#include <sys/types.h>
+#include <sys/stat.h>

 #include "uftrace.h"
 #include "utils/perf.h"
```
In addition, please see the following logs.
```
$ uftrace record t-fork

$ ls uftrace.data/perf-cpu*
uftrace.data/perf-cpu23.dat  uftrace.data/perf-cpu2.dat  uftrace.data/perf-cpu4.dat
```
As you can see above, it has only `perf-cpu2.dat`, `perf-cpu4.dat`, and `perf-cpu23.dat` and it looks good. But the following dump message shows that it has `perf-cpu0.dat`, `perf-cpu1.dat`, and `perf-cpu2.dat`.
```
$ uftrace dump
    ...
reading perf-cpu0.dat
21418602.297706444   3672: [event] linux:sched-in(200001)
21418602.298691132   3672: [event] linux:task-exit(200005)
reading perf-cpu1.dat
21418602.293363071   3661: [event] linux:task-name(200006)
21418602.297639296   3672: [event] linux:task-new(200004)
21418602.297712261   3661: [event] linux:sched-out(200002)
reading perf-cpu2.dat
21418602.298814014   3661: [event] linux:sched-in(200001)
21418602.299298193   3661: [event] linux:task-exit(200005)
```
Could you please fix this as well?

**Comment by honggyukim** [comment]
What about this?
```diff
diff --git a/cmds/dump.c b/cmds/dump.c
index a951443..c8d17cd 100644
--- a/cmds/dump.c
+++ b/cmds/dump.c
@@ -782,7 +782,7 @@ static void dump_raw_perf_start(struct uftrace_dump_ops *ops,
        if (cpu == 0)
                pr_out("\n");

-       pr_out("reading perf-cpu%d.dat\n", cpu);
+       pr_out("reading %s\n", perf->perf_name);

        raw->file_offset = 0;
 }
diff --git a/utils/perf.c b/utils/perf.c
index 9d5e086..79f984d 100644
--- a/utils/perf.c
+++ b/utils/perf.c
@@ -277,6 +277,7 @@ int setup_perf_data(struct uftrace_data *handle)
                perf[i].fp = fopen(globbuf.gl_pathv[i], "r");
                if (perf[i].fp == NULL)
                        pr_err("open failed: %s", globbuf.gl_pathv[i]);
+               perf[i].perf_name = xstrdup(globbuf.gl_pathv[i]);
        }

        handle->nr_perf = globbuf.gl_pathc;
diff --git a/utils/perf.h b/utils/perf.h
index 1c4de16..f97b5da 100644
--- a/utils/perf.h
+++ b/utils/perf.h
@@ -108,6 +108,7 @@ static inline void record_perf_data(struct uftrace_perf_writer *perf,

 struct uftrace_perf_reader {
        FILE                    *fp;
+       char                    *perf_name;
        bool                    valid;
        bool                    done;
        int                     type;
```
`perf_name` should be free, but didn't include it here yet.

**Comment by honggyukim** [comment]
The similar change is needed for `kernel-cpuXX.dat` as well.

**Comment by yeominstall** [comment]
@honggyukim Thank you for your advice. I applied it.

File: utils/perf.c [hunk 1]
```diff
@@ -133,13 +133,28 @@ int setup_perf_record(struct uftrace_perf_writer *perf, int nr_cpu, int pid,
...
-		if (perf->fp[cpu])
+		if (perf->fp[cpu]) {
+			f_no = fileno(perf->fp[cpu]);
+			sprintf(proclnk, "/proc/self/fd/%d", f_no);
+			r = readlink(proclnk, filename, PATH_MAX);
+			if (r < 0) {
+				pr_warn("failed to find perf record: %m\n");
+				continue;
+			}
+			filename[r] = '\0';
```
Comment (namhyung) [inline utils/perf.c]: I think you'd better adding `output_dir` to `uftrace_perf_writer` and making filenames using cpu index.

File: utils/perf.c [hunk 2]
```diff
@@ -263,7 +265,7 @@ int setup_perf_data(struct uftrace_data *handle)
 	size_t i;
 	int ret = -1;
 
-	xasprintf(&pattern, "%s/perf-cpu*.dat", handle->dirname);
+	xasprintf(&pattern, "perf-cpu*.dat");
```
Comment (namhyung) [inline utils/perf.c]: Why is this change needed?  And I think it changes behavior.  Did you test this?

File: utils/perf.c [hunk 3]
```diff
@@ -277,6 +279,7 @@ int setup_perf_data(struct uftrace_data *handle)
 		perf[i].fp = fopen(globbuf.gl_pathv[i], "r");
 		if (perf[i].fp == NULL)
 			pr_err("open failed: %s", globbuf.gl_pathv[i]);
+		perf[i].name = xstrdup(globbuf.gl_pathv[i]);
```
Comment (namhyung) [inline utils/perf.c]: I'm not sure the ordering is correct (like perf-cpu23 vs. perf-cpu4)

Comment (honggyukim) [reply utils/perf.c]: I know that `perf-cpu23` comes first than `perf-cpu4`. It's a trivial problem, but looks tricky to handle.

File: utils/perf.c (hunk 1 above)
Comment (yeominstall) [reply utils/perf.c]: Did you mean below?
```cpp
struct uftrace_perf_writer {
	int			*event_fd;
	char			*output_dir;
	void			**page;
...
```

File: utils/perf.c (hunk 2 above)
Comment (yeominstall) [reply utils/perf.c]: Oh, It seems to make some problem. I'll put it back.

The reason why I changed this line is, the former format `"%s/perf-cpu*.dat"` makes string result like below:
```
reading .//perf-cpu1.dat
2588.701366131   2489: [event] linux:task-name(200006)
2588.704225677   2489: [event] linux:sched-out(200002)
...
```
But as you already pointed it out, I think it can be a matter when user try to print information on the external directory. 
I will trim `globbuf.gl_pathv[i]`. Thank you.

File: utils/perf.c (hunk 1 above)
Comment (namhyung) [reply utils/perf.c]: yep

Comment (yeominstall) [reply utils/perf.c]: I think that will be better. Thank you for the comment.

File: utils/perf.c (hunk 3 above)
Comment (yeominstall) [reply utils/perf.c]: Yes, as @honggyukim said earlier, it is not a big problem and is related with linux filesystem directory ordering. But I agree that it needs to be ordered properly. I think it can be the one of future work. How do you think about that?

File: utils/perf.c (hunk 1 above)
Comment (namhyung) [reply utils/perf.c]: Then you can simple stat the file without calling `fileno() + readlink()`:

```c
for (cpu = 0; cpu < perf->nr_events; cpu++) {
   ...
   snprintf(filename, sizeof(filename), "%s/perf-cpu%d.dat", perf->output_dir, cpu);
   if (!stat(filename, &stbuf) && stbuf.st_size == 0)
      remove(filename);
}
```

File: utils/perf.c (hunk 3 above)
Comment (namhyung) [reply utils/perf.c]: Hmm.. did you test it with kernel data?  I think it might not work..

Comment (honggyukim) [reply utils/perf.c]: What about working on this only for `perf-cpuXX` data in this PR? It doesn't have to make it too much complicated to kernel data as of now.

Comment (namhyung) [reply utils/perf.c]: I'm ok with that.

**Comment by namhyung** [comment]
Due to potential issues, I won't accept this change.  I think it's ok to have some empty files..  If you really want to fix this, please do with more testing, thanks.
